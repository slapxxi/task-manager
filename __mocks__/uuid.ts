const uuidMock = {
  counter: 0,
  __reset() {
    this.counter = 0;
  },
  v4() {
    const id = `unique-id-${this.counter}`;
    this.counter += 1;
    return id;
  },
};

export default uuidMock;
