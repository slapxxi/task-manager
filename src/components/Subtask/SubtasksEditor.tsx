import { createSubtask } from '@lib';
import { Subtask as ISubtask } from '@local/types';
import React, { useRef } from 'react';
import Subtask from './Subtask';
import Subtasks from './Subtasks';

interface Props {
  subtasks: ISubtask[];
  focused?: boolean;
  isEmpty?: boolean;
  onCreate?: (subtask: ISubtask) => void;
  onEdit?: (subtask: ISubtask) => void;
  onRemove?: (subtask: ISubtask) => void;
}

function SubtasksEditor({
  subtasks,
  focused,
  isEmpty,
  onCreate,
  onEdit,
  onRemove,
}: Props) {
  const lastFocusedRef = useRef<number>(NaN);

  function handleEdit(subtask: ISubtask) {
    if (onEdit) {
      onEdit(subtask);
    }
  }

  function handleCreate(subtask: ISubtask) {
    if (onCreate) {
      onCreate(createSubtask(subtask));
    }
  }

  function handleAdd() {
    if (onCreate) {
      onCreate(createSubtask({}));
      lastFocusedRef.current = subtasks.length;
    }
  }

  function handleRemove(subtask: ISubtask) {
    if (onRemove) {
      if (subtasks.length === 1) {
        lastFocusedRef.current = 0;
      } else {
        lastFocusedRef.current -= 1;
      }
      onRemove(subtask);
    }
  }

  function handleFocus(index: number) {
    lastFocusedRef.current = index;
  }

  function handleBlur() {
    lastFocusedRef.current = NaN;
  }

  return (
    <div>
      <Subtasks
        subtasks={isEmpty ? [...subtasks, createSubtask({})] : subtasks}
        render={({ subtask, index }) =>
          isEmpty && index === subtasks.length ? (
            <Subtask subtask={subtask} onEdit={handleCreate} />
          ) : (
            <Subtask
              focus={focused && lastFocusedRef.current === index}
              subtask={subtask}
              onEdit={handleEdit}
              onFocus={() => handleFocus(index)}
              onBlur={handleBlur}
              onRemove={handleRemove}
              onSubmit={handleAdd}
            />
          )
        }
      />
    </div>
  );
}

export default React.memo(SubtasksEditor);
