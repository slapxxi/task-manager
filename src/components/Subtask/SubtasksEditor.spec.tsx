import { createSubtask, Keys } from '@lib';
import React from 'react';
import { fireEvent, render } from 'react-testing-library';
import uuid from 'uuid';
import SubtasksEditor from './SubtasksEditor';

window.requestAnimationFrame = (fn) => setTimeout(fn, 10);

const subtasks = [
  { id: 'first', description: 'first', completed: false },
  { id: 'second', description: 'second', completed: false },
];

beforeEach(() => {
  // @ts-ignore
  uuid.__reset();
});

it('renders', () => {
  const { container } = render(<SubtasksEditor subtasks={subtasks} />);
  expect(container.firstChild).toMatchSnapshot();
});

it('invokes callback when subtask created', () => {
  const spy = jest.fn();
  const { getByTestId } = render(<SubtasksEditor subtasks={[]} onCreate={spy} isEmpty />);
  fireEvent.change(getByTestId('input'), { target: { value: 'new' } });
  expect(spy).toHaveBeenCalledWith({
    id: 'unique-id-0',
    description: 'new',
    completed: false,
  });
});

it('invokes callback when subtask submitted', () => {
  const spy = jest.fn();
  const { getByTestId } = render(
    <SubtasksEditor
      subtasks={[createSubtask({ id: 'test', description: 'existing' })]}
      onCreate={spy}
    />,
  );
  fireEvent.keyDown(getByTestId('input'), {
    keyCode: Keys.enter,
  });
  expect(spy).toHaveBeenCalledWith({
    id: 'unique-id-0',
    description: '',
    completed: false,
  });
});
