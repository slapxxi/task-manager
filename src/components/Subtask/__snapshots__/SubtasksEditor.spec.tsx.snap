// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<div>
  <ul
    class="subtasks"
  >
    <li
      class="subtask"
    >
      <div
        class="container"
      >
        <div
          class="checkbox"
        >
          <div
            style="position: relative; display: flex; align-items: center; height: 15px;"
          >
            <input
              class="input"
              type="checkbox"
            />
            <svg
              clip-rule="evenodd"
              height="15"
              preserveAspectRatio="none"
              viewBox="0 0 100 100"
              width="15"
            >
              <circle
                cx="50"
                cy="50"
                fill="none"
                r="44"
                stroke="#99a"
                stroke-width="8"
              />
            </svg>
          </div>
        </div>
        <input
          class="sc-bwzfXH PGmuR"
          data-testid="input"
          placeholder="What should be done..."
          type="text"
          value="first"
        />
        <svg
          class="trashbin"
          height="16"
          viewBox="viewBox"
          width="16"
        >
          <use
            id="icon-id"
            xlink:href="#id"
          />
        </svg>
      </div>
    </li>
    <li
      class="subtask"
    >
      <div
        class="container"
      >
        <div
          class="checkbox"
        >
          <div
            style="position: relative; display: flex; align-items: center; height: 15px;"
          >
            <input
              class="input"
              type="checkbox"
            />
            <svg
              clip-rule="evenodd"
              height="15"
              preserveAspectRatio="none"
              viewBox="0 0 100 100"
              width="15"
            >
              <circle
                cx="50"
                cy="50"
                fill="none"
                r="44"
                stroke="#99a"
                stroke-width="8"
              />
            </svg>
          </div>
        </div>
        <input
          class="sc-bwzfXH PGmuR"
          data-testid="input"
          placeholder="What should be done..."
          type="text"
          value="second"
        />
        <svg
          class="trashbin"
          height="16"
          viewBox="viewBox"
          width="16"
        >
          <use
            id="icon-id"
            xlink:href="#id"
          />
        </svg>
      </div>
    </li>
  </ul>
</div>
`;
